ceno_zkvm/src/witness.rs

```diff
@@ -28,97 +26,102 @@ use crate::{
+                .collect(),
             num_col,
+            padding_strategy,
         }
     }
 
+    pub fn num_padding_instances(&self) -> usize {
+        next_pow2_instance_padding(self.num_instances()) - self.num_instances()
+    }
+
     pub fn num_instances(&self) -> usize {
-        self.values.len() / self.num_col - self.num_padding_rows
+        self.values.len() / self.num_col
     }
 
-    pub fn num_padding_instances(&self) -> usize {
-        self.num_padding_rows
+    pub fn row(&self, row: usize) -> &[T] {
+        &self.values[row * self.num_col..(row + 1) * self.num_col]
     }
 
-    pub fn iter_rows(&self) -> Chunks<MaybeUninit<T>> {
+    pub fn iter_rows(&self) -> Chunks<T> {
         self.values.chunks(self.num_col)
     }
 
-    pub fn iter_mut(&mut self) -> ChunksMut<MaybeUninit<T>> {
+    pub fn iter_mut(&mut self) -> ChunksMut<T> {
         self.values.chunks_mut(self.num_col)
     }
 
-    pub fn par_iter_mut(&mut self) -> rayon::slice::ChunksMut<MaybeUninit<T>> {
+    pub fn par_iter_mut(&mut self) -> rayon::slice::ChunksMut<T> {
         self.values.par_chunks_mut(self.num_col)
     }
 
-    pub fn par_batch_iter_mut(
-        &mut self,
-        num_rows: usize,
-    ) -> rayon::slice::ChunksMut<MaybeUninit<T>> {
+    pub fn par_batch_iter_mut(&mut self, num_rows: usize) -> rayon::slice::ChunksMut<T> {
         self.values.par_chunks_mut(num_rows * self.num_col)
     }
+}
 
-    pub fn par_batch_iter_padding_mut(
-        &mut self,
-        num_rows: usize,
-    ) -> rayon::slice::ChunksMut<'_, MaybeUninit<T>> {
-        let valid_instance = self.num_instances();
-        self.values[valid_instance * self.num_col..]
-            .as_mut()
-            .par_chunks_mut(num_rows * self.num_col)
-    }
-
-    pub fn de_interleaving(mut self) -> Vec<Vec<T>> {
-        (0..self.num_col)
-            .map(|i| {
-                self.values
-                    .par_iter_mut()
-                    .skip(i)
-                    .step_by(self.num_col)
-                    .map(|v| unsafe { mem::replace(v, mem::MaybeUninit::uninit()).assume_init() })
-                    .collect::<Vec<T>>()
-            })
-            .collect()
+impl<F: Field + From<u64>> RowMajorMatrix<F> {
+    // Returns column number `column`, padded appropriately according to the stored strategy
+    pub fn column_padded(&self, column: usize) -> Vec<F> {
+        let num_instances = self.num_instances();
+        let num_padding_instances = self.num_padding_instances();
+        let last_element = self.row(num_instances - 1)[column];
+
+        let padding_iter = (num_instances..num_instances + num_padding_instances).map(|i| {
+            match &self.padding_strategy {
+                InstancePaddingStrategy::Custom(fun) => F::from(fun(i as u64, column as u64)),
+                InstancePaddingStrategy::RepeatLast => last_element,
+                _ => F::ZERO,
+            }
+        });
+
+        self.values
+            .iter()
```

Comment:
"self.values" including non-padding + padding. the iteration here will go through padding part, and the chaining of padding_iter will make padding occur twice.

we can fix here
- self.values range should be [0..num_instances * num_col] for non-default (non zero) padding
- for padding strategy as default (zero), we can skip chaining with padding_iter(), as vector now is initialized deterministics.

Reply:
Just clarified per offline discussion, self.values only contain non-padding value 👍